src/arraymancer/nn_dsl.nim

```diff
@@ -2,11 +2,406 @@
+    error("Layer body must be a statement list: \"" & $toStrLit(sectionInfo.body) & "\"", sectionInfo.body)
+
+  for layer in sectionInfo.body:
+
+    if layer.kind != nnkCall or
+    layer.len != 2 or
+    layer[0].kind != nnkIdent or
+    layer[1].kind != nnkStmtList or
+    layer[1].len != 1 or
+    layer[1][0].kind != nnkCall or
+    layer[1][0].len < 1 or
+    layer[1][0][0].kind != nnkIdent:
+      error("Unknown configuration of layer section: \"" & $toStrLit(layer) & "\"", layer)
+
+    result.add LayerInfo(
+      name: layer[0],
+      typeName: layer[1][0][0]
+    )
+    if layer[1][0].len >= 2:
+      result[^1].arguments = layer[1][0][1..^1]
+
+proc createModelType(layerInfos: seq[LayerInfo], modelName: NimNode): NimNode =
+
+  # creates the type defintion of the model, e.g.:
+  #  type
+  #    SomeConvNet[T] = object
+  #      cv: Conv2D[T]
+  #      mp: Maxpool2D[T]
+  #      fl: Flatten[T]
+
+  let underlyingTypeSymbol = genSym(nskType, "T")
+  var recList = newNimNode(nnkRecList)
+  for layerInfo in layerInfos:
+    doAssert layerInfo.name.kind == nnkIdent
+    doAssert layerInfo.typeName.kind == nnkIdent
+    recList.add newIdentDefs(
+      layerInfo.name,
+      newNimNode(nnkBracketExpr).add(
+        layerInfo.typeName,
+        underlyingTypeSymbol
+      )
+    )
+  
+  doAssert modelName.kind == nnkIdent
+
+  result = newNimNode(nnkTypeSection).add(
+    newNimNode(nnkTypeDef).add(
+      modelName,
+      newNimNode(nnkGenericParams).add(
+        newIdentDefs(
+          underlyingTypeSymbol,
+          newEmptyNode()
+        )
+      ),
+      newNimNode(nnkObjectTy).add(
+        newEmptyNode(),
+        newEmptyNode(),
+        recList
+      )
+    )
+  )
+  
+proc createInitProc(layerInfos: seq[LayerInfo], sectionInfo: SectionInfo, modelName: NimNode): NimNode =
+
+  # creates init function of the model, e.g.:
+  #   proc init[T](ctx: Context[AnyTensor[T]], modelType: typedesc[SomeConvNet[T]], h: auto; w: auto): SomeConvNet[T] =
+  #     template cv(): auto =
+  #       result.cv
+  #     template mp(): auto =
+  #       result.mp
+  #     template fl(): auto =
+  #       result.fl
+  #     cv = init(ctx, Conv2D[T], @[1, h, w], 20, (5, 5))
+  #     mp = init(ctx, Maxpool2D[T], cv1.outShape, (2, 2), (0, 0), (2, 2))
+  #     fl = init(ctx, Flatten[T], mp.outShape)
+
+  doAssert modelName.kind == nnkIdent
+
+  var body = newNimNode(nnkStmtList)
+  for layerInfo in layerInfos:
+    body.add(
+      newNimNode(nnkTemplateDef).add(
+        layerInfo.name,
+        newEmptyNode(),
+        newEmptyNode(),
+        newNimNode(nnkFormalParams).add ident"auto",
+        newEmptyNode(),
+        newEmptyNode(),
+        newStmtList(
+          newDotExpr(
+            ident"result",
+            layerInfo.name
+          )
+        )
+      )
+    )
+
+  let
+    ctxSymbol = genSym(nskParam, "ctx")
+    underlyingTypeSymbol = ident($toStrLit(genSym(nskGenericParam, "T")))# TODO fix this
```

Comment:
Why the back and forth conversion from an `nnkSym` back to a `nnkIdent`? Does a symbol cause problems here?